sim/warrior/charge.go

```diff
@@ -6,28 +6,82 @@ import (
 	"github.com/wowsims/cata/sim/core"
 )
 
+func performCharge(sim *core.Simulation, warrior *Warrior, metrics *core.ResourceMetrics, rage float64) {
+	warrior.AddRage(sim, rage, metrics)
+	if warrior.DistanceFromTarget > core.MaxMeleeRange {
+		warrior.PseudoStats.MovementSpeedMultiplier *= 2
+		warrior.MoveTo(core.MaxMeleeRange-1, sim) // movement aura is discretized in 1 yard intervals, so need to overshoot to guarantee melee range
+	} else {
+		warrior.ChargeAura.Deactivate(sim)
+	}
+}
+
 func (warrior *Warrior) RegisterCharge() {
 	metrics := warrior.NewRageMetrics(core.ActionID{SpellID: 100})
+	rage := float64(15 + 5*warrior.Talents.Blitz)
+	chargeMinRange := 8.0
+
+	warrior.ChargeAura = warrior.RegisterAura(core.Aura{
+		Label:    "Charge",
+		ActionID: core.ActionID{SpellID: 100},
+		Duration: 5 * time.Second,
+	})
+
+	warrior.RegisterMovementCallback(func(sim *core.Simulation, position float64, kind core.MovementUpdateType) {
+		if warrior.ChargeAura.IsActive() && kind == core.MovementEnd {
+			if warrior.DistanceFromTarget >= chargeMinRange {
+				// Has moved out, charge back in
+				performCharge(sim, warrior, metrics, rage)
+			} else {
+				// Has charged back in, reset movement speed
+				warrior.PseudoStats.MovementSpeedMultiplier /= 2
+				warrior.ChargeAura.Deactivate(sim)
```

Comment:
It's cleaner to put the movement speed changes within the `OnGain` and `OnExpire` callbacks for the Charge buff aura directly.